pkg/volume/empty_dir/quota.go

```diff
@@ -0,0 +1,183 @@
+	return "", stderr.String(), err
+}
+
+// Apply sets the actual quota on a device for an emptyDir volume if possible. Will return an error
+// if anything goes wrong during the process. (not an XFS filesystem, etc) If the volume medium is set
+// to memory, or no FSGroup is provided (indicating the request matched an SCC set to RunAsAny), this
+// method will effectively no-op.
+func (xqa *xfsQuotaApplicator) Apply(dir string, medium api.StorageMedium, pod *api.Pod, fsGroup *int64, quota resource.Quantity) error {
+
+	if medium == api.StorageMediumMemory {
+		glog.V(3).Infof("Skipping quota application due to memory storage medium.")
+		return nil
+	}
+	isXFS, err := xqa.isXFS(dir)
+	if err != nil {
+		return err
+	}
+	if !isXFS {
+		return fmt.Errorf("unable to apply quota: %s is not on an XFS filesystem", dir)
+	}
+	if fsGroup == nil {
+		// This indicates the operation matched an SCC with FSGroup strategy RunAsAny.
+		// Not an error condition.
+		glog.V(3).Infof("Unable to apply XFS quota, no FSGroup specified.")
+		return nil
+	}
+
+	volDevice, err := xqa.getFsDevice(dir)
+	if err != nil {
+		return err
+	}
+
+	err = xqa.applyQuota(volDevice, quota, *fsGroup)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// isXFS checks if the empty dir is on an XFS filesystem.
+func (xqa *xfsQuotaApplicator) isXFS(dir string) (bool, error) {
+	out, _, err := xqa.cmdRunner.RunFsTypeCommand(dir)
+	if err != nil {
+		return false, fmt.Errorf("unable to check filesystem type for emptydir volume %s: %s", dir, err)
+	}
+	if strings.Contains(out, "xfs") {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (xqa *xfsQuotaApplicator) applyQuota(volDevice string, quota resource.Quantity, fsGroupID int64) error {
+	_, stderr, err := xqa.cmdRunner.RunApplyQuotaCommand(volDevice, quota, fsGroupID)
+	if err != nil {
+		glog.Errorf("Error running xfs_quota: %s", err)
+		return err
+	}
+	// xfs_quota is very happy to fail but return a success code, likely due to its
+	// interactive shell approach. Grab stderr, if we see anything written to it we'll
+	// consider this an error.
+	if len(stderr) > 0 {
+		glog.Error("Error running xfs_quota: ", stderr)
+		return errors.New("xfs_quota wrote to stderr")
+	}
+
+	glog.V(3).Infof("XFS quota applied: device=%s, quota=%d, fsGroup=%d", volDevice, quota.Value(), fsGroupID)
+	return nil
+}
+
+// getFsDevice returns the fs device for a given emptydir path. To do this we
+// run df on the emptyDir path, returning a header line and the line we're
+// interested in. The first string token in that line will be the device name.
+func (xqa *xfsQuotaApplicator) getFsDevice(dir string) (string, error) {
+	out, _, err := xqa.cmdRunner.RunFsDeviceCommand(dir)
+	if err != nil {
+		return "", fmt.Errorf("unable to find filesystem device for emptyDir volume %s: %s", dir, err)
+	}
+	fsDevice, parseErr := ParseFsDevice(out)
+	return fsDevice, parseErr
+}
+
+func ParseFsDevice(dfOutput string) (string, error) {
```

Comment:
> Hmm, this was so I could use it in my extended test, but looking at it I probably should expose something higher level.

I'm not requiring a refactor, just some explicit docs that warn people off from using it would satisfy me.